codes/1963_cpp_wodnjs.cpp

```diff
@@ -0,0 +1,112 @@
+#include<iostream>
+#include<queue>
+#include<utility>
+
+using namespace std;
+bool not_prime[10000];
+int vis[10000];
+void prime_checker();
+int bfs(int from, int to);
+void clear_vis();
+int main() {
+	ios::sync_with_stdio(false);
+	cin.tie(nullptr);
+	cout.tie(nullptr);
+	int t_case;
+	cin >> t_case;
+	prime_checker();
+	for (int t = 0; t < t_case; t++)
+	{
+		int from, to;
+		cin >> from >> to;
+		int bfs_ret=bfs(from, to);
+		if(bfs_ret==-1)
+			cout << "Impossible\n";
+		else
+			cout << vis[to] - 1 << '\n';
+	}
+	return 0;
+}
+void prime_checker() {
+	for (int i = 2; i *i < 10000; i++)
+	{
+		if (not_prime[i] == false) 
+		{
+			for (int j = 2 * i; j < 10000; j += i)
+			{
+				not_prime[j] = true;
+			}
+		}
+	}
+}
+void clear_vis() {
+	for (int i = 1000; i < 10000; i++)
+	{
+		vis[i] = 0;
+	}
+}
+int bfs(int from, int to) {
+	int from_arr[4];
+	queue<int> q;
+	clear_vis();
+	vis[from] = 1;
+	q.push(from);
+	while (!q.empty())
+	{
+		int front_num = q.front();
+		for (int i = 3; i > -1; i--)
+		{
+			from_arr[i] = q.front() % 10;
+			q.front() /= 10;
+		}
+		q.pop();
+	
+		if (front_num== to) 
+			return 0;
+		for (int i = 1; i < 10; i++)//맨 앞자리 수는 0이 올 수 없다.
+		{
+			if (from_arr[0] == i) 
+				continue;
+			int cmp = i * 1000 + from_arr[1] * 100 + from_arr[2] * 10 + from_arr[3];
+			if (!not_prime[cmp]&&vis[cmp]==0)
+			{
+				q.push(cmp);
+				vis[cmp] = vis[front_num] + 1;
+			}
+		}
+		for (int i = 0; i < 10; i++)
+		{
+			if (from_arr[1] == i) 
+				continue;
+			int cmp = from_arr[0] * 1000 + i * 100 + from_arr[2] * 10 + from_arr[3];
+			if (!not_prime[cmp] && vis[cmp] == 0)
+			{
+				q.push(cmp);
+				vis[cmp] = vis[front_num] + 1;
+			}
+		}
+		for (int i = 0; i < 10; i++)
+		{
+			if (from_arr[2] == i) 
+				continue;
+			int cmp = from_arr[0] * 1000 + from_arr[1] * 100 + i * 10 + from_arr[3];
+			if (!not_prime[cmp] && vis[cmp] == 0)
+			{
+				q.push(cmp);
+				vis[cmp] = vis[front_num] + 1;
+			}
+		}
+		for (int i = 0; i < 10; i++)
+		{
+			if (from_arr[3] == i) 
+				continue;
+			int cmp = from_arr[0] * 1000 + from_arr[1] * 100 + from_arr[2] * 10 + i;
+			if (!not_prime[cmp] && vis[cmp] == 0)
+			{
+				q.push(cmp);
+				vis[cmp] = vis[front_num] + 1;
+			}
+		}
```

Comment:
자리수별로 for문을 돌리기 보다
for(자릿수)
   for(바꿀 수)
이렇게 짜주세요

Reply:
자리 수마다 로직이 조금씩 다른데 일반화 할 수 있는 방법이 있을까요?

Reply:
맨 앞자리 수가 0이 아니여야한다는 것이라면 조건문에 1000이상의 수라는 것을 추가하거나 1000미만의 수에 대해서 vis를 -1와 같은 수를 채워 넣으면 됩니다.

간략하게 코드 예시를 보여드리자면 현재 숫자를 num이라 하면
```suggestion
		for(int digit = 1; digit <= 1000; digit *= 10){
		    int digit_ num = (num / digit) % 10;
		    for(int i=0; i <=9; i++){
		        if(digit_num == i)
		            continue;
		        int next_num = num - digit_num * digit + i * digit;
```